### PR TITLE
Migrate to draft 07 and add ajv Debug Server

### PR DESCRIPTION
**Pull request by iwootten**
### PR Context

Currently, errors returned from validator are not usually very helpful.

We'd like to make use of some of the features of the json schema draft 07 specification - mainly the polymorphism features described [here](https://json-schema.org/draft-07/json-schema-validation.html#rfc.section.6.6) in order to get better error messages from validator when schemas fail validation.

This work originally started as making a decision on which language it was best to pursue continued development of validator within, as it seemed ajv (written in nodejs) was the better option. Unfortunately, ajv does not support polymorphism in the way we originally understood meaning that we have the opposite problem of many errors being returned (as errors from every block type are considered) and this makes interpretation of them difficult. ajv has an open [issue](https://github.com/epoberezkin/ajv/issues/1119) to support a "discriminator" keyword which should allow us to reduce the number of messages returned. ajv is the preferred option - in like for like json schema tests of our own schemas from runner it reduced validation times from around 2mins to about 15s on my hardware.

Therefore this PR has become an update of validator to support draft 07, along with the introduction of an ajv based server to be used during debug of schemas. We could split the schema/questionnaire validation between two implementations, but this would make our current workflow even more complex. The errors returned by the ajv implementation are ordered by their path length, which should mean the first error is the most useful as it will have matched on a particular block.

- [x] Update schemas to draft 07.
- [x] Update to latest version of jsonschema library. 

### How to Review

See if you agree with the decisions made. Check tests pass.

### CODE REVIEW

**Comment by iwootten** [comment]
[Evaluate json schema validator libraries](https://trello.com/c/bh25hvHt/3498-evaluate-json-schema-validator-libraries)

File: README.md [hunk 1]
```diff
@@ -54,3 +54,25 @@ Run the following to format all json files in the schemas directory:
...
+The node implementation depends on yarn. Once installed install ajv using:
+
+```
+yarn
+```
+
+Run the ajv based version of validator from within the ajv/ directory.
+
+```
+DEBUG=validator node app.js
```
Comment (ajmaddaford) [inline README.md]: What's the significance of the `DEBUG` setting?

File: .python-version [hunk 2]
```diff
@@ -1 +1 @@
-3.6.3
+3.7.2
```
Comment (ajmaddaford) [inline .python-version]: Any reason not to upgrade to 3.8.0?

Comment (iwootten) [reply .python-version]: Only that we're currently using this version on runner. Using 3.8 would save the travis install step.

File: README.md (hunk 1 above)
Comment (iwootten) [reply README.md]: codacy doesn't like using console.log statements - express suggests using the 'debug' module and it defines what it's logging out as part of the require call.

File: schemas/common_definitions.json [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
-  },
+  "$id": "https://eq.ons.gov.uk/common_definitions.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
   "q_code": {
     "type": "string",
     "description": "A question code used by downstream systems to identify answers.",
     "pattern": "^[0-9a-z]+$"
   },
-  "definitions": {
+  "idx": {
```
[the remark this answers is not in the document]
Comment (iwootten) [reply schemas/common_definitions.json]: I'm open to suggestions

File: schemas/answers/checkbox.json [hunk 4]
```diff
@@ -1,10 +1,11 @@
 {
-  "$schema": "http://json-schema.org/draft-04/schema",
+  "$id": "https://eq.ons.gov.uk/answers/checkbox.json",
+  "$schema": "http://json-schema.org/draft-07/schema",
   "answer": {
     "type": "object",
     "properties": {
       "id": {
-        "$ref": "../common_definitions.json#/id"
+        "$ref": "../common_definitions.json#/idx"
```
[the remark this answers is not in the document]
Comment (iwootten) [reply schemas/answers/checkbox.json]: I could only get certain refs to work as urls with ajv.

File: tests/test_questionnaire_id_regex.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import json
+
+from app.validation.validator import Validator
+
+validator = Validator()
+
+
+def create_schema_with_id(schema_id):
```
[the remark this answers is not in the document]
Comment (iwootten) [reply tests/test_questionnaire_id_regex.py]: Fair enough. Hadn't noticed that when consolidating from the original tests.

File: schemas/answers/checkbox.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajmaddaford) [reply schemas/answers/checkbox.json]: I just tried changing a few and they seem to work, which ones cause problems?

File: schemas/common_definitions.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajmaddaford) [reply schemas/common_definitions.json]: identifier?

[the remark this answers is not in the document]
Comment (ajmaddaford) [reply schemas/common_definitions.json]: If we nested all definitions under a `definitions` property, we could use id.  Use of `definitions` is documented as 'customary (but not required)' at https://json-schema.org/understanding-json-schema/structuring.html.

[the remark this answers is not in the document]
Comment (ajmaddaford) [reply schemas/common_definitions.json]: I've also seen `$defs` used in other schemas (https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.2.5), but I got an ajv error when using id nested under it.

[the remark this answers is not in the document]
Comment (iwootten) [reply schemas/common_definitions.json]: I've updated to use "identifier"

File: README.md (hunk 1 above)
Comment (MebinAbraham) [inline README.md]: We instruct how to install the python version required using pyenv.  We should do the same for node with nvm.

File: ajv/app.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  const validate = ajv.compile(require('../schemas/questionnaire_v1.json'))
+
+  app.post('/validate', (req, res, next) => {
+    const valid = validate(req.body)
+    debug('Validating questionnaire: ' + req.body.title)
+    if (!valid) {
+      return res.json({
+        success: false,
+        errors: validate.errors.sort((a, b) => {
```
Comment (MebinAbraham) [inline ajv/app.js]: Can we use something clearer than a, b?

File: README.md [hunk 7]
```diff
@@ -39,18 +39,48 @@ Then, in another terminal window/tab, navigate to a checked out copy of eq-surve
...
+## Validating with ajv
+
+Also included is a node based version of the json schema validation which may be used during development to assist with
+debugging errors. This returns more errors than we'd currently like due to the way polymorphism works for each of our
+blocks.
+
+The node implementation depends on yarn. Once installed install ajv using:
+
+```
+yarn
```
Comment (MebinAbraham) [inline README.md]: This is now duplicated.

**Pull request merged**